analytics_data_api/v0/tests/views/test_learners.py

```diff
@@ -404,9 +412,192 @@ def test_pagination(self):
         ({'course_id': 'edX/DemoX/Demo_Course', 'page_size': 101}, 'illegal_parameter_values'),
+        ({'course_id': 'edX/DemoX/Demo_Course', 'segments': 'a_non_existent_segment'}, 'illegal_parameter_values'),
+        ({'course_id': 'edX/DemoX/Demo_Course', 'ignore_segments': 'a_non_existent_segment'},
+         'illegal_parameter_values'),
     )
     @ddt.unpack
     def test_bad_request(self, parameters, expected_error_code):
         response = self.authenticated_get('/api/v0/learners/', parameters)
         self.assertEqual(response.status_code, 400)
         self.assertEqual(json.loads(response.content)['error_code'], expected_error_code)
+
+
+@ddt.ddt
+class CourseLearnerMetadataTests(DemoCourseMixin, VerifyCourseIdMixin,
+                                 LearnerAPITestMixin, TestCaseWithAuthentication):
+    """
+    Tests for the course learner metadata endpoint.
+    """
+
+    def _get(self, course_id):
+        """Helper to send a GET request to the API."""
+        return self.authenticated_get('/api/v0/course_learner_metadata/{}/'.format(course_id))
+
+    def test_no_course_id(self):
+        response = self.authenticated_get('/api/v0/course_learner_metadata/')
+        self.assertEqual(response.status_code, 404)
+
+    @ddt.data(
+        {},
+        {'highly_engaged': 1},
+        {'disengaging': 1},
+        {'struggling': 1},
+        {'inactive': 1},
+        {'unenrolled': 1},
+        {'highly_engaged': 3, 'disengaging': 1},
+        {'disengaging': 10, 'inactive': 12},
+        {'highly_engaged': 1, 'disengaging': 2, 'struggling': 3, 'inactive': 4, 'unenrolled': 5},
+    )
+    def test_segments_unique_learners(self, segments):
+        """
+        Tests segment counts when each learner belongs to at most one segment.
+        """
+        learners = [
+            {'username': '{}_{}'.format(segment, i), 'course_id': self.course_id, 'segments': [segment]}
+            for segment, count in segments.items()
+            for i in xrange(count)
+        ]
+        self.create_learners(learners)
+        expected = self._get_full_engagement_ranges()
+        response = self._get(self.course_id)
+        self.assertEqual(response.status_code, 200)
+        expected_segments = {"highly_engaged": 0, "disengaging": 0, "struggling": 0, "inactive": 0, "unenrolled": 0}
+        expected_segments.update(segments)
+        expected['segments'] = expected_segments
+        self.assertDictContainsSubset(expected, json.loads(response.content))
+
+    def test_segments_same_learner(self):
+        """
+        Tests segment counts when each learner belongs to multiple segments.
+        """
+        self.create_learners([
+            {'username': 'user_1', 'course_id': self.course_id, 'segments': ['struggling', 'disengaging']},
+            {'username': 'user_2', 'course_id': self.course_id, 'segments': ['disengaging']}
+        ])
+        expected = self._get_full_engagement_ranges()
```

Comment:
I think this could use some cleanup since we both touched these test cases. I could do it or you could, but I'll hold off for now in case `self._get_full_engagement_ranges` changes.